adoptingerlang.org

```diff
@@ -3444,44 +3446,74 @@ initContainers:
-
-The only necessary thing to do for graceful shutdown is to delay =init:stop()= from running until the =Pod= is removed from the =kube-proxy= =Endpoint=. The process can wait up to 30 seconds before Kubernetes sends a =SIGKILL= signal that will force the process to terminate.
-
-In =service_discovery= the =readinessProbe= is on a separate path, =/ready=, and with a higher frequency, every 5 seconds.
+**** Readiness, Liveness and Startup Probes
+
+Each container in a =Pod= can define a =readiness=, =liveness= and =startup= probe. The
+=liveness= probe . A startup probe can be added to delay when the =liveness= and
+=readiness= probes start being checked.
+
+Probes are defined either as a command to execute in the container, as an HTTP
+=GET= request made to a given port and path, or as an attempt to open a TCP
+connection to a specific port.
+
+A =livenessProbe= is not required to ensure restart of failed containers. If the
+process in the container crashes or stops for any reason, it will already be
+restarted. The probe should only be used for cases the process does not crash
+but for some reason has deadlocked and can not respond to requests. For this
+reason if a =livenessProbe= is included it should be kept extremely simple. It is
+more often than not safer to leave out the =livenessProbe= entirely and rely on
+the node crashes, metric based alerting and auto-scaling.
+
+One part of auto-scaling is based on the =readinessProbe=. The =readinessProbe=
+tells Kubernetes the =Pod= is ready to receive traffic. When a =Pod= becomes =unready=
+because a container is failing the =readinessProbe=, it is removed from the
+=Service= resulting in traffic being directed at the existing =Pods=, increasing
+their load, then a new =Pod= will be started.
+
+Not using the =livenessProbe= for this purpose means the containers are still
+available, since they aren't killed, and the reason for failing the check can be
+examined. This may mean attaching a shell if it isn't completely frozen or
+forcing a crash dump to be written. In a very resource constrained environment
+you may prefer to have the =livenessProbe= in order to have deadlocked containers
+and their =Pods= cleaned up right away, instead of waiting for manual intervention
+and the auto-scaler to scale down after load falls.
+
+Another use for the =readinessProbe= is if the application has to periodically do work
+where you don't want to be handling requests, or have to do some sort of
+maintenance, it can return a =503= status response for readiness causing it to be
+removed from the =Service= backends.
+
+Nothing special is needed to have a graceful shutdown, for any reason, whether
```

Comment:
Eeeeh, if you don't remove yourself from the load balancer pool, some gateways (like ALBs or ELBs) may still try to send traffic to a process that's halfway through shutting down, and that will cause interrupted connections that are forwarded to the client. This may however be stack dependent because not all stacks clean up as nicely and orderly as an OTP tree.

Reply:
Removal should already be handled by k8s. There is always delay, even in the case of the readinessProbe being setup to return 503 on shutdown since the check is only done every `period` seconds, but its handled by k8s and not the application itself is my point. Maybe it needs rewording?

Reply:
Yeah that could make it clearer. It's possible this was just my misunderstanding and I didn't realize that being in a shutting down state also automatically took you out.

Reply:
Actually, I should probably mention the use of `prep_stop`. I haven't implemented this in `service_discovery` because it is a little tricky in its case since it has 2 applications, one for http and one for grpc, that both have to wait until active requests are handled before shutting down.

I suppose I should also mention the need to have a proper boot order and supervision tree so that stuff like the database pool aren't shutdown before the http app can pause to complete open requests.

I could just add a "see Graceful Shutdown section (to come)" for more...

Reply:
Eh, maybe it isn't tricky, was overthinking it. It is the same whether they are both in the release or not. Each can wait for their active connections one after the other and the SIGKIll takes care of it if they take too long.

Reply:
Wait, yea, this is just k8s specific. I will just add a callout to another section and with saying in k8s you don't need to do anything special, but see this section for how to handle the shutdown in your Erlang application.